colossalai/gemini/stateful_tensor_container.py

```diff
@@ -0,0 +1,131 @@
+    evicted, meaning the type of its device is CUDA, the other is that it isn't pinned in CUDA
+    memory, meaning its state isn't COMPUTE.
+
+    This container should get a stateful tensor when it become HOLD_LIKE from COMPUTE.
+    And it pops stateful tensors in function, `evict_tensors`.
+
+    In order to acquire an optimal eviction policy, users may need to offer computation step
+    index of each stateful tensor. So we can use a heap to maintain all potential evictable
+    statefule tensors. When poping, we can get the stateful tensor that used furthest in
+    current computation step.
+    """
+
+    def __init__(self, compute_step_dict: Dict[StatefulTensor, List[int]], total_step: int):
+        self.compute_step_dict = compute_step_dict
+        self.total_step = total_step
+
+    @abstractmethod
+    def empty(self) -> bool:
+        pass
+
+    @abstractmethod
+    def create(self, stateful_tensor_list: List[StatefulTensor]) -> None:
+        pass
+
+    @abstractmethod
+    def push(self, stateful_tensor: StatefulTensor, cur_step: int) -> None:
+        pass
+
+    @abstractmethod
+    def pop(self) -> Optional[StatefulTensor]:
+        pass
+
+
+class QueueSTContainer(BaseSTContainer):
+    """Queue type stateful tensor container. This is used in 'cpu' tensor placement policy.
+    It pops potential evictable stateful tensors in FIFO.
+    """
+
+    def __init__(self, compute_step_dict: Dict[StatefulTensor, List[int]], total_step: int):
+        super().__init__(compute_step_dict, total_step)
+        self.container = None
+
+    def empty(self) -> bool:
+        assert self.container is not None
+        return self.container.empty()
+
+    def create(self, stateful_tensor_list: List[StatefulTensor]) -> None:
+        self.container = queue.SimpleQueue()
+        for stateful_tensor in stateful_tensor_list:
+            self.container.put(stateful_tensor)
+
+    def push(self, stateful_tensor: StatefulTensor, cur_step: int) -> None:
+        self.container.put(stateful_tensor)
+
+    def pop(self) -> Optional[StatefulTensor]:
+        ret = None
+        while not self.empty():
+            out_tensor = self.container.get()
+            if evict_check(out_tensor):
+                ret = out_tensor
+                break
+
+        return ret
+
+
+class HeapSTContainer(BaseSTContainer):
+    """Heap type stateful tensor container. This is used in 'auto' tensor placement policy.
+    It pops potential evictable stateful tensors in the order of the distance between current
+    step and next used step.
+    """
+
+    def __init__(self, compute_step_dict: Dict[StatefulTensor, List[int]], total_step: int):
+        super().__init__(compute_step_dict, total_step)
+        self.container = None
+
+    def empty(self) -> bool:
+        assert self.container is not None
+        return self.container == []
+
+    def create(self, stateful_tensor_list: List[StatefulTensor]) -> None:

```

Comment:
Can you try to use a PQ?
from queue import PriorityQueue

Reply:
Heap is faster.
It has an linear time creation, while PQ needs nlogn time.
Furthermore, PQ is the same as heap, since we provide our API like a queue.